agent/agent.go

```diff
@@ -179,6 +182,11 @@ func (r *Agent) Run(ctx context.Context, recipe recipe.Recipe) (run Run) {
 
 	// start listening.
 	// this process is blocking
+	defer func() {
```

Comment:
this still seems incorrect. from the place where the sink is setup (assuming line 146 where `r.setupSink()` is being called) to this line, there are multiple `return` in-between. what happens when we end-up returning early?

correct way to do this is usually to have the `defer Close()` at the site of initialisation of this resource (after error checking)

```golang
sink, err := r.setupSink(...)
if err != nil {
   return err
}
defer sink.Close()
```

Reply:
@spy16 thanks for this, will take note for future usecase as well